File: leetcode/go/567.permutation-in-string.go
func checkInclusion(s1 string, s2 string) bool {
    if len(s1) > len(s2) {
        return false
    }

    s1Counts := make([]int, 26)
    for _, c := range s1 {
        s1Counts[c-'a']++
    }

    s2Counts := make([]int, 26)
    l := len(s1)

    for i := 0; i < l; i++ {
        s2Counts[s2[i]-'a']++
    }

    for i := l; i < len(s2); i++ {
        if equal(s1Counts, s2Counts) {
            return true
        }

        s2Counts[s2[i]-'a']++
        s2Counts[s2[i-l]-'a']--
    }

    if equal(s1Counts, s2Counts) {
        return true
    }

    return false
}

func equal(a, b []int) bool {
    for i := 0; i < len(a); i++ {
        if a[i] != b[i] {
            return false
        }
    }
    return true
}

File: leetcode/cpp/567.permutation-in-string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        if (s1.length() > s2.length()) {
            return false;
        }

        vector<int> s1Counts(26, 0);
        for (char c : s1) {
            s1Counts[c - 'a']++;
        }

        vector<int> s2Counts(26, 0);
        int l = s1.length();

        for (int i = 0; i < l; ++i) {
            s2Counts[s2[i] - 'a']++;
        }

        for (int i = l; i < s2.length(); ++i) {
            if (equal(s1Counts, s2Counts)) {
                return true;
            }

            s2Counts[s2[i] - 'a']++;
            s2Counts[s2[i - l] - 'a']--;
        }

        if (equal(s1Counts, s2Counts)) {
            return true;
        }

        return false;
    }
private:
    bool equal(const vector<int>& a, const vector<int>& b) {
        for (int i = 0; i < a.size(); ++i) {
            if (a[i] != b[i]) {
                return false;
            }
        }
        return true;
    }
};

